### DISCUSSION TITLE
How do we "Fit()" the terminal when terminal is initially hidden?

### DISCUSSION THREAD

**Discussion by TomzBench** [sourcelair/xterm.js#4578] [Q&A]
It would be helpful to know how the fit addon works, Because I am having trouble getting the terminal to size correctly when the terminal is initially loaded with `display: none`.  The terminal exists in my UI as part of a tabbed pane, which is not rendered on initial view.  However, a user can click the terminal tab and the terminal will enter the view.

On the initial view of the terminal tab, the terminal overflows the height significantly. However, tabbing away and back again will then render correctly.  Im not sure why that happens.  I also use ReactResizable component and can resize the terminal.  This part works great however, because I can use `onResize` events to `fit()` the terminal.

So my only problem is when the terminal loads initially in a hidden state, it seems to be unable to detect the heights correctly.  So knowing how the fitaddon works a bit could help me figure out the best way to load my terminal into view. I am doing my best to try and call `fit()` when the terminal is tabbed into view but this has not been successful so far.

The docs on the fit addon are very light, it could be helpful to go over some gotchas or limitations in the fit Addon

**Comment by jerch** [comment]
You can find the addon code here: https://github.com/xtermjs/xterm.js/blob/master/addons/xterm-addon-fit/src/FitAddon.ts

**Reply by TomzBench** [reply]
Thanks.

My initially hidden terminal has 0 for dimensions from the `_renderService`.  But after a tab navigation puts the terminal in view I call fit again.  At this moment, the `parentElements` height is appropriately calculated.  But the `_renderService` still claims 0 dimensions and so fit returns `undefined` when initially loaded into view.  `dims` seems correct the second time around however and so re-navigating back works like normal..

**Reply by TomzBench** [reply]
I can get it to work when i call `resize` with some arbitrary values and then call `fit`.

To make this more sensible, i gave my React Terminal component an imperative handle with a method called `unhide`. which will internally will resize and then fit, so not to call `resize` every time, and only when not hidden.

Not sure if this could be a bug in the fitAddon, which does not handle uninitialized `cols/rows` correctly.  I think instead of returning `undefined` when the rows/cols are unknown, you can resize and then fit
